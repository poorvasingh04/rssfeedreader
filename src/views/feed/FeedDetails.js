import { useRoute } from '@react-navigation/core';
import React, { useEffect } from 'react';
import {
  FlatList,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import SCREENS from '../../constants/SCREENS';
import { getFeedDetails } from '../../services/actions';
import FeedItem from './feed item/FeedItem';

function FeedDetails({navigation}) {

  const feed = useRoute().params.feed || {};
  const { id, name } = feed;
  const dispatch = useDispatch();

  const feedsDetails = useSelector(state => state.FeedDetails.data);

  useEffect(() => {
    let cancel = false;
    const runEffect = async () => {      
      if (cancel) {
        return;
      }
     
      if(feedsDetails) {
        navigation.setOptions({
          title: `${name} (${feedsDetails.length})`,
        });
      }

      dispatch(getFeedDetails(id));
    };
    runEffect();
  
    return () => {
      cancel = true;
    }
  }, [feedsDetails]);

  openItem = (item) => {
    const { url, title } = item;

    if(navigation) navigation.navigate(
      SCREENS.FEED_ITEM_DETAILS,
      {
        url: url,
        title,
      }
    );
  }

  renderItem = ({ item }) => <FeedItem item={item} openItem={openItem} />

  return (
    <FlatList
      keyExtractor={(item) => item.id}
      renderItem={renderItem}
      data={feedsDetails}
    />
  );
}

export default FeedDetails;