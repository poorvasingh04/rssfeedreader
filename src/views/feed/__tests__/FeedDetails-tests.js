import 'react-native';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import FeedDetails from '../FeedDetails';
import { FlatList, Text, TouchableOpacity } from 'react-native';

jest.mock('react-native-webview', () => 'WebView');

jest.mock('@react-navigation/core', () => ({
  useRoute: () => {
    return {
      params: {
        feed: { id: '2', name: 'BBC'},
      }
    }
  },
 }));

describe('FeedDetails test cases-----', () => {
  const firstFeed = {
    "id": "1",
    "title": "Myanmar protests: Protesters killed in bloody Yangon crackdown",
    "description": "At least 14 protesters have been killed in clashes in Myanmar's main city Yangon as politicians ousted by the military coup called for \"revolution\".\nSecurity forces opened fire in the Yangon area of Hlaing Tharyar as protesters used sticks and knives....",
    "date": "03-14-2021 7:11 pm",
    "url": "https://www.bbc.com/news/world-asia-56395085"
  };

  const mockStore = configureStore()
  let store, wrapper;
  const navigateAction = jest.fn();
  const navigation = {
    navigate: (props, data) => navigateAction(props, data),
  }

  beforeEach(() => {
    store = mockStore({
      FeedDetails: {
        data: [firstFeed],
      }
    });
    wrapper = renderer.create(
      <Provider store={store}>
        <FeedDetails navigation={navigation} />
      </Provider>
    );
  });

  it('renders correctly', () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('renders correct number of components', () => {
    const testInstance = wrapper.root;
    expect(testInstance.findByType(FlatList).props.data.length).toBe(1);
  });

  it('test render item', () => {
    const testInstance = wrapper.root;
    expect(testInstance.findAllByType(TouchableOpacity).length).toBe(1);
    expect(testInstance.findAllByType(Text).length).toBe(4);
  });

  it('should open feed details on tap', () => {
    const feedCard = wrapper.root.findAllByType(TouchableOpacity)[0];
    feedCard.props.onPress();
    expect(navigateAction).toBeCalledWith('FeedItem Details', {
      url: firstFeed.url,
      title: firstFeed.title,
    });
  });
  
})