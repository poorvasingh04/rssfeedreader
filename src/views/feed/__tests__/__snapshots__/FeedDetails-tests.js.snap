// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedDetails test cases----- renders correctly 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "date": "03-14-2021 7:11 pm",
        "description": "At least 14 protesters have been killed in clashes in Myanmar's main city Yangon as politicians ousted by the military coup called for \\"revolution\\".
Security forces opened fire in the Yangon area of Hlaing Tharyar as protesters used sticks and knives....",
        "id": "1",
        "title": "Myanmar protests: Protesters killed in bloody Yangon crackdown",
        "url": "https://www.bbc.com/news/world-asia-56395085",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 10,
            "elevation": 4,
            "margin": 10,
            "opacity": 1,
            "padding": 20,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 2,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#000000",
              "fontSize": 18,
              "fontWeight": "700",
            }
          }
        >
          Sunday, March 14, 2021 7:11 PM
        </Text>
        <Text
          style={
            Object {
              "color": "#000000",
              "fontSize": 16,
              "fontWeight": "500",
              "marginBottom": 5,
              "marginTop": 5,
            }
          }
        >
          Myanmar protests: Protesters killed in bloody Yangon crackdown
        </Text>
        <Text>
          At least 14 protesters have been killed in clashes in Myanmar's main city Yangon as politicians ousted by the military coup called for "revolution".
Security forces opened fire in the Yangon area of Hlaing Tharyar as protesters used sticks and knives....
        </Text>
        <Text
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignSelf": "flex-start",
                "borderColor": "#000000",
                "borderRadius": 4,
                "borderWidth": 0.5,
                "color": "#000000",
                "fontSize": 14,
                "marginTop": 10,
                "overflow": "hidden",
                "padding": 10,
              },
              Object {
                "backgroundColor": "#FFFFFF",
              },
            ]
          }
        >
          Mark as favorite
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
